simulated mouse fixes, rightclick, dblclick

kuceb opened the pull request:
### mouse events
mouse events added in this PR:
```diff
+ pointerup
+ pointerdown
+ pointermove
+ pointerover
+ pointerenter
+ pointerleave
+ pointerout
mouseup
mousedown
+ mousemove
+ mouseover
+ mouseenter
+ mouseleave
+ mouseout
click
dblclick
+ contextmenu
```
- [x] send mouseover event during click commands 
fix #2928 
fix #1847
fix #1016
- [x] send pointer events (`pointerdown`, `pointerup`, `pointermove`) during click commands 
fix #2957
fix #2045
- [x] send mousemove events during click commands 
fix #2521 
- [x] add mouseout events

- [x] fix incorrect click coordinates inside [same domain] iframe
fix #4347

### right click, double click
- [x] allow `dblclick` command to accept same options + perform actions similar to `click`
fix #116 
fix #3798
fix #2768
fix #3224
- [x] add "rightclick" command
fix #53 

### trigger
- [x] add `screenX`/`screenY` properties to events sent with `.trigger`
fix #3187 
  - [x] tests

#### tasks
- [x] pointer events
- [x] console log shows all events
- [x] fix dblclick
- [x] right click
- [x] ~send `selectionchange` event on `document`~ _browser already does this even when out of focus_
- [x] add test for clicking inside iframe
- [x] fix coords incorrect when clicking inside iframe
- [x] waiting on fix for #4396 
- [x] typedefs
- [x] fix messed up console table of events
- [x] add test for cy.click checking checkboxes
- [x] add comment for element-moving-during-click algorithm
- [x] refactor action commands to reuse click logic
- [x] add `fromAutWindow` positional coords for red hitbox
- [x] add test for iframe red hitbox coords
 
#### non-blocking tasks
- [ ] write new doc for .rightclick() command https://github.com/cypress-io/cypress-documentation/issues/2102
- [ ] update document events simulated in `.click()` https://docs.cypress.io/api/commands/click.html#Events https://github.com/cypress-io/cypress-documentation/issues/2173
- [ ] update `.dblclick()` doc `options` to match `.click()` options. https://docs.cypress.io/api/commands/dblclick.html#Arguments
- [ ] kitchensink update https://github.com/cypress-io/cypress-example-kitchensink/pull/298

### Guide for review

- review console messages during `cy.click, cy.dblclick, and cy.rightclick`
- make sure tests cover all use cases of `dblclick` and `rightclick`
- let me know if I missed an issue this fixes

### User facing changes:
#### API:
- add `cy.rightclick()`

#### typedefs changes:
- add `cy.rightclick()` typedef

#### UX:
- console messages when pinning `click` command has more information:
![image](https://user-images.githubusercontent.com/14625260/65705605-c8ab7580-e056-11e9-9c67-5c4b32fd355d.png)
warning messaging in table:
![image](https://user-images.githubusercontent.com/14625260/65706196-ee854a00-e057-11e9-822a-6b7996275ebe.png)
no mouse move events needed:
![image](https://user-images.githubusercontent.com/14625260/65729934-18f1fa00-e08e-11e9-8df3-17e169cab0ea.png)

Bkucera commented:
WIP going back on for dblclick and rightclick changes

jennifer-shehane commented:
Changed the base to `develop`, is this still WIP since there's a checkbox not checked?

Bkucera commented:
It's in a reviewable state, but I do still have one task to do

kuceb committed:
rename mouse/keyboard

kuceb committed:
apply changes on this branch with rename

kuceb committed:
temp 07/18/19 [skip ci] type_spec minor changes

kuceb committed:
cleanup type_spec, allow unused-vars for args

kuceb committed:
fix missing return

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 07/19/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 07/22/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 07/22/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
remove accidental spec, use beforeEach visit in click_spec

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
add tests for cy.trigger changes

kuceb committed:
re-run build

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
add typedefs

kuceb committed:
temp 08/08/19 [skip ci]

kuceb committed:
fix type_spec, click_spec

brian-mann commented:
I see some cleanup but overall this is excellent.

kuceb committed:
fix console table event table logging

kuceb committed:
Merge branch 'develop' into issue-2957

kuceb committed:
fix spec utils

kuceb committed:
fix invalid clicking-into-iframe spec

brian-mann reviewed (dismissed):
- these are mostly notes, but adding them now. some of which i will internally handle - so don't jump into these yet.
- i'm currently going through each of the issues and finding the tests and running them to understand and gauge the solutions we implemented.
- also i'm looking for our collective notes on the algorithm changes so we can surface those and them not get lost
- we talked internally about using `while` loops vs tail-call optimized recursions. no change needed but some research needs to be done to see which pattern we should employ. there are places that we traverse the dom using recursion which may not be tail-call optimized and could blow out the stack. however research suggests recursion is more performant that while loops. probably is easier to convert those recursive calls to `while` loops, but ideally we could look into this further and then decide on a general strategy of how/when to use one vs the other.

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/click_spec.js:
```diff
@@ -257,25 +294,37 @@ describe('src/cy/commands/actions/click', () => {
-        expect(events).to.deep.eq(['mouseup', 'click'])
+      cy.get('#button').click().should('not.have.focus')
+
+      cy.getAll('$btn', 'pointerdown mousedown pointerup mouseup click').each(shouldBeCalled)
```
need to think about a potentially better way to do this. not necessarily bad or wrong just need to discuss

Bkucera commented on packages/driver/src/cy/mouse.js:
```diff
@@ -1,143 +1,653 @@
 const $dom = require('../dom')
+const $elements = require('../dom/elements')
+const $ = require('jquery')
```
we have not. there is a plugin but it has strange bugs sometimes where it moves all your requires to the bottom of the file

Bkucera commented on packages/driver/src/cy/mouse.js:
```diff
@@ -1,143 +1,653 @@
+        which: 1,
+        buttons: 1,
+        detail: 0,
+        pressure: 0.5,
```
yes, added comment to clarify

Bkucera commented on packages/driver/src/cy/mouse.js:
```diff
@@ -1,143 +1,653 @@
+  const _eventModifiers = $Keyboard.fromModifierEventOptions(evtOptions)
+  const modifiers = $Keyboard.modifiersToString(_eventModifiers)
+
+  const evt = new constructor(evtName, _.extend({}, evtOptions, { bubbles, cancelable }))
```
:+1:

kuceb committed:
address review, cleanup

Bkucera commented on packages/driver/src/cy/mouse.js:
```diff
@@ -1,143 +1,653 @@
+const sendMouseEvent = (el, evtOptions, evtName, bubbles = false, cancelable = false) => {
+  // IE doesn't have event constructors, so you should use document.createEvent('mouseevent')
+  // https://dom.spec.whatwg.org/#dom-document-createevent
+  const constructor = el.ownerDocument.defaultView.MouseEvent
```
:+1:

Bkucera commented on packages/driver/src/cy/mouse.js:
```diff
@@ -1,143 +1,653 @@
-      const modifiers = keyboard.getActiveModifiersArray()
+      const el = doc.elementFromPoint(x, y)
+
+      // mouse._mouseMoveEvents(el, { x, y })
```
:+1:

Bkucera commented on packages/driver/src/cy/mouse.js:
```diff
@@ -1,143 +1,653 @@
+const _ = require('lodash')
+const $Keyboard = require('./keyboard')
+const $selection = require('../dom/selection')
+const debug = require('debug')('driver:mouse')
```
:+1:

Bkucera commented on packages/driver/src/cy/keyboard.js:
```diff
@@ -37,6 +38,34 @@ const initialModifiers = {
+
+const modifiersToString = (modifiers) => {
+  return _.keys(
+    _.pickBy(modifiers, (val) => {
```
:+1:

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

brian-mann commented on packages/driver/test/cypress/integration/commands/actions/click_spec.js:
```diff
@@ -257,25 +294,37 @@ describe('src/cy/commands/actions/click', () => {
-        expect(events).to.deep.eq(['mouseup', 'click'])
+      cy.get('#button').click().should('not.have.focus')
+
+      cy.getAll('$btn', 'pointerdown mousedown pointerup mouseup click').each(shouldBeCalled)
```
This was an old comment, disregard

brian-mann reviewed (changes requested):
Can you write out the event firing algorithm including how we handle things like when the element moves in response to an event? 

It will take some legwork, and it may possibly be better diagramed as a flowchart (rather than nested bullet points) - i don't necessarily care, i just want it documented somewhere. I would prefer if its documented in the code itself so it doesn't become stale - and new engineers will be able to pick it up. I'd also accept a link to an external document in the code if necessary. We can also use flowchart.js to accomplish both - where the algorithm is written in code in notes but a physical flowchart asset is generated from the code. That way it never gets stale.

as you described they are incoherent. We need to build institutional knowledge around these algorithms and make them very transparent. Jennifer created one awhile ago for the visibility algorithm - she'll link you here in a sec.

jennifer-shehane commented:
maybe flowchart in code similar visibility algo 

![nC0TrUIVMOLJObyg-F2826 copy](https://user-images.githubusercontent.com/1271364/65627302-195e9800-df9d-11e9-8657-960f39ff91e5.png)

Bkucera commented:
@jennifer-shehane ok, do we have the flowchart-as-code stuff documented anywhere

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
temp 09/25/19 [skip ci]

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
add test for clicking checkbox, cleanup click_spec

kuceb committed:
document mouse click and mouse move algo, adjust mouse move, add test for recursive movement

brian-mann committed:
Merge branch 'develop' into issue-2957

brian-mann committed:
cleanup / formatting

brian-mann committed:
scope the getCommandLogWithText to the active runnable to prevent leakage

brian-mann committed:
cleanup tests, make values static so tests pass consistently

- use attach listener alias APIs
- remove magic number constants
- fix some incoherent html / script tag stuff

brian-mann committed:
begin implementation for returning fromAutWindow

- restore original fromWindow algorithm
- begin writing fromAutWindow for calculating red hitbox coordinates
which are layered on the AUT

brian-mann reviewed (changes requested):
update test for iframe coordinates to actually have coords
- fromAutWindow
- remove the $elements.getFromDocCoords(x, y, win)
- write a proper test for the iframe coordinates specifically for drawing the red hitbox
- remove docCoords

brian-mann committed:
move describe block with the others

brian-mann committed:
refactor actions to DRY up duplicated logic between click/dblclick/rightclick

- tighten up the mouse.js method names, name them consistently

brian-mann committed:
remove comment

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

kuceb committed:
fix iframe coords and test

jennifer-shehane committed:
Merge branch 'develop' into issue-2957

kuceb committed:
rename fromDocCoords->fromWindowCoords, fix type_spec ux test

kuceb committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

kuceb committed:
Merge remote-tracking branch 'origin/develop' into issue-2957

brian-mann committed:
tighten up method, remove unnecessary arg

brian-mann committed:
Merge remote-tracking branch 'origin/issue-2957' into issue-2957

brian-mann committed:
add debug logic for retries to console.error() non cypress errors

brian-mann committed:
renamed fromWindow -> fromElWindow, and fromViewport -> fromElViewport

brian-mann committed:
handle calculating the fromElWindow coordinates inline

brian-mann committed:
extract out function for clarity, receive consistent arg names

brian-mann committed:
cleanup, add lots of comments

brian-mann committed:
Merge branch 'develop' into issue-2957

# Conflicts:
#	packages/driver/src/cy/commands/screenshot.coffee
#	packages/driver/test/cypress/integration/dom/coordinates_spec.coffee
#	packages/driver/test/cypress/integration/e2e/dom_hitbox.spec.js
